crates/loader/src/local.rs

```diff
@@ -485,6 +485,20 @@ impl LocalLoader {
         dest_root: &Path,
         exclude_files: &[String],
     ) -> Result<()> {
+        if glob_or_path == ".." || glob_or_path.ends_with("/..") {
+            bail!("A file pattern can't end in a parent directory path (..)\nIf you want to include a directory, use source-destination form, or a glob pattern ending in **/*.\nLearn more: https://spinframework.dev/writing-apps#including-files-with-components");
+        }
+        if glob_or_path == "." || glob_or_path.ends_with("/.") {
+            bail!("A file pattern can't end in a current directory path (.)\nIf you want to include a directory, use source-destination form, or a glob pattern ending in **/*.\nLearn more: https://spinframework.dev/writing-apps#including-files-with-components");
+        }
+
+        if glob_or_path == "*" {
+            terminal::warn!("A component is including the entire application directory as asset files. This is unlikely to be what you want.\nIf this is what you want, use the pattern \"./*\" to avoid this warning.\nLearn more: https://spinframework.dev/writing-apps#including-files-with-components\n");
+        }
+        if glob_or_path == "**/*" {
+            terminal::warn!("A component is including the entire application directory tree as asset files. This is unlikely to be what you want.\nIf this is what you want, use the pattern \"./**/*\" to avoid this warning.\nLearn more: https://spinframework.dev/writing-apps#including-files-with-components\n");
+        }
```

Comment:
I wonder if these warnings should be moved to somewhere more `spin up`-specific. I believe the containerd shim shares this code.

Reply:
Here perhaps? https://github.com/spinframework/spin/blob/72500b233fabfaafa41e34a0a61eee6892079d86/src/commands/up.rs#L490-L497

Reply:
Thinking down the rabbit hole, an alternative would be to integrate `tracing` with `terminal` so that we can emit warnings anywhere we like (annotated with some appropriate field like `show_this_to_the_user_please = true` and translate those into `terminal` output. That's a lot of churn to maintain conceptual purity; what do you think?

Reply:
Should anyone doubt my zealotry for conceptual purity, I have in the past done a "state monad" in JavaScript just to have a way to pass warnings back to the correct level.  I will not be saying anything further about the 4-year gap in my resume.

I guess the tracing approach is do-able, but it's then not clear to me how a host (shim or whatever) chooses what warnings to present or not present - or is the idea that "in `spin up`, warnings are off by default, except for ones labelled as 'actually this one is not off by default'"?  This sounds like it could be more broadly useful anyway.

Reply:
> "in `spin up`, warnings are off by default, except for ones labelled as 'actually this one is not off by default'"?

Yes, something like that. See also https://github.com/spinframework/spin/issues/3144#issuecomment-2901213109